### PR TITLE
Line numbering for code block

### PR DESCRIPTION
**Pull request by yipinliu**
Add line numbering for code block like markdown.

### Example
```typ
#set raw(
    lineno: true
)

= Line numbering
```=rs
fn main() {
    println!("Hello, world!");
}
` ``

```=typ
#let count = 10
#let nums = range(1, count + 1)
#let fib(n) = (
  if n <= 2 { 1 }
  else { fib(n - 1) + fib(n - 2) }
)
` ``
```

image

Closes: #344

### CODE REVIEW

**Comment by laurmaedje** [comment]
I would prefer to not have syntax for this and to just add a property that can be configured with a set rule.

**Comment by yipinliu** [comment]
> I would prefer to not have syntax for this and to just add a property that can be configured with a set rule.

@laurmaedje I updated the implementation. And what do you think of the problem @rktjmp mentioned above? I have no idea where the line wrapping happens and don't know how to deal with it.:worried:

**Comment by yipinliu** [comment]
image

Should we align all the line numbers in the code blocks of the entire document?

**Comment by 0ip** [comment]
@yipinliu  Could you make the delimiter character configurable/optional?

**Comment by yipinliu** [comment]
image

@0ip Done.

**Comment by 0ip** [comment]
@yipinliu Thank you ☃️

File: library/src/text/raw.rs [hunk 1]
```diff
@@ -255,6 +316,15 @@ fn to_syn(RgbaColor { r, g, b, a }: RgbaColor) -> synt::Color {
...
 }
 
+fn num_width(mut num: usize) -> usize {
+    let mut width = 0;
+    while num != 0 {
+        width += 1;
+        num /= 10;
+    }
+    width
+}
```
Comment (istudyatuni) [inline library/src/text/raw.rs:326]:
```suggestion
fn num_width(num: usize) -> usize {
    (num.checked_ilog10().unwrap_or_default() + 1) as usize
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c8474458172314fa97326b703a0dcb92

**Comment by LaurenzV** [comment]
I personally don't like that the delimiter is a separate property. How about instead defining a settable `numbering` property that takes either a boolean or a function? It would then behave as follows (which also seems a bit more consistent to me when considering how it's done for [lists](https://typst.app/docs/reference/layout/enum/)

`#set raw(numbering: false)` -> Default, no line numbers at all
`#set raw(numbering: true)` -> Default numbering (without separators or anything) starting at 1
`#set raw(numbering: n => [#n|])` -> Shows the numbering with a custom formatting (you can add a separator, but you can do customize it whatever way you want really. For example, if you want it to start at 0 you could just do `#set raw(numbering: n => {#n - 1}`

Would something speak against implementing it this way?

**Comment by istudyatuni** [comment]
> either a boolean or a function

Not sure about boolean, maybe use just `none` for disabling, and function for customizing?

**Comment by laurmaedje** [comment]
I agree with both previous comments. This should be `numbering: Option<Numbering>`, that is, allow `none`, `"1"`, and a function. The default way to enable it would be just `"1"`, consistent with all other numberings in Typst. The linebreaking problem could be resolved by displaying the code in a grid with two columns and n rows where n is the number lines. That would also make proper alignment of the numbers simple.

**Comment by yipinliu** [comment]
image

It seems that the grid's column can't be set to right-aligned.

**Comment by LaurenzV** [comment]
Works fine if you do it like this:
````
#set page(width: 180pt)
#set text(6pt)
// Line numbering of code block
#set raw(
     numbering: el => [#align(right)[#el]]
)
```rs
fn main() {
    println!("Hello, world!");








    return;
}
```
````

**Comment by laurmaedje** [comment]
You could make it right aligned by default by calling `.aligned(...)` on all the pieces of content that go in the left column.

**Comment by yipinliu** [comment]
image

File: library/src/text/raw.rs [hunk 2]
```diff
@@ -127,75 +142,208 @@ impl Synthesize for RawElem {
...
         let foreground = THEME
             .settings
             .foreground
             .map(to_typst)
             .map_or(Color::BLACK, Color::from)
             .into();
 
-        let mut realized = if matches!(lang.as_deref(), Some("typ" | "typst" | "typc")) {
+        let mut realized = if let Some(numbering) = numbering.as_ref() {
+            realize_with_numbering(text, lang, foreground, vt, styles, numbering)?
```
Comment (laurmaedje) [inline library/src/text/raw.rs:157]: I'm not a fan of having two completely separate code paths for this.

**Review by laurmaedje — CHANGES_REQUESTED** [review]

File: library/src/text/raw.rs (hunk 2 above)
Comment (yipinliu) [reply library/src/text/raw.rs]: Well, the logic of these two is very different, and I couldn't find a good way to merge them.

File: tests/typ/text/raw-code.typ [hunk 3]
```diff
@@ -71,3 +71,27 @@ int main() {
...
+fn main() {
+    println!("Hello, world!");
+
+
+
+
+
+
+
+
```
[the remark this answers is not in the document]
Comment (yipinliu) [reply tests/typ/text/raw-code.typ:91]: Do you mean like the following?
```rs
fn main() {
    println!("Hello, world!");\n\n\n\n\n\n\n\n\n\n
    return; 
}\n
```

**Comment by laurmaedje** [comment]
I think this feature requires more design work regarding API and implementation to make it nice and flexible, so I am going to close this PR for now.

**Pull request closed without merging**